## MEDLINE Papers Network Analysis

Every year, MELINE(Medical Literature Analysis and Retrieval System Online) would publish a sample of acamdemic papers covering the life sciences and medcine. 

You can find the MEDLINE dataset here: [ftp://ftp.nlm.nih.gov/nlmdata/sample/medline/](ftp://ftp.nlm.nih.gov/nlmdata/sample/medline/). There are eight separate XML files with the total size of 928m. 

In this notebook, I'm going to investigate three topics about the network analysis:

[Part1: Community Detection](#Part1: Community Detection)

Considering the languages used, which group of users is more likely to publish papers together? Is this true that the smaller the language group is, the higher the clustering effect would occur?

[Part2: Co-occurence of Attributes](#Part2: Co-occurence of Attributes)



Considering the descriptions and keywords of the papers, what combinations of descriptions or keywords are more likely to appear together?

[Part3: Link Prediction](#Part3: Link Prediction)

Although this is not a classic social network, can we build a model to predict waht combinations of authors are more likely to publish a paper together(having an edge)?

Dependencies: 

networkX. Pure python implementation of graph algorithms, works fine with this dataset, but if you are analyzing bigger graphs, the scalability could be a problem. iGraph is not better but GraphX would be a good alternative, coming with Pregel API is even better.

H2O. Very fast data processing and model training thanks to its structure of compressing the data and store it on the JVM heap. The documentaion is complex but it is promising and offers H2O with hadoop/Spark to scale.

### Preperation

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib import style
style.use('ggplot')
% matplotlib inline

# name your own file path
dirpath = "/Users/apple/Downloads/datasets/aas/medline/"

In [2]:
import xml.etree.cElementTree as ET
import codecs

# use iterparse to extract info from xml files
def extract_papers(fs):
    '''
    parameter: xml files
    return: dictionary of papers
            key: PMID
            value: list of authors
                   list of language used
                   list of descriptions
                   list of keywords
    '''
    papers = {}
    for f in fs:
        with codecs.open(f, 'r') as fi:
            path = []
            for event, elem in ET.iterparse(fi, events=("start", "end")):
                if event == "start":
                    path.append(elem.tag)
                    if elem.tag == 'PMID':
                        k = elem.text
                        papers[k] = {'auth':[], 'lang':[], 'description':[], 'keyword':[]}
                    elif 'Author' in path:
                        if elem.tag == 'LastName':
                            papers[k]['auth'].append([elem.text])
                        if elem.tag == 'ForeName' and elem.text is not None:
                            papers[k]['auth'][-1].append(elem.text)
                    elif elem.tag == 'Language' and elem.text is not None:
                        langs = elem.text
                        if len(langs)%3 != 0:
                            print '{0}: incorrect/no languages -> {1}'.format(k, langs)
                        else:
                            for i in range(len(langs)/3):
                                papers[k]['lang'].append(langs[3*i:(3*i+3)])
                    elif elem.tag == 'DescriptorName':
                        papers[k]['description'].append(elem.text)
                    elif elem.tag == 'Keyword':
                        papers[k]['keyword'].append(elem.text)
                elif event == "end":
                    path.pop()
                elem.clear()
    for k,v in papers.iteritems():
        if len(v['auth']) > 0:
            papers[k]['auth'] = [' '.join(auth) for auth in papers[k]['auth'] if auth[0] is not None]
    return {k:v for k,v in papers.iteritems() if len(papers[k]['auth']) > 0}


files = []
for i in ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h']:
    fn = dirpath+'medsamp2016'+i+'.xml'
    files.append(fn)
    
papers = extract_papers(files)

In [3]:
# take a look at the result
# filter the result with more than one language used
[(k,v) for k,v in papers.iteritems() if len(v['lang']) > 1][:2]

[('23515729',
  {'auth': ["Mont'Alverne",
    'Galdino Lara Maia',
    'Pinheiro Marcela Cunha',
    u'Levy C\xedntia Souto',
    'Vasconcelos Glauber Gean de',
    u'Souza Neto Jo\xe3o David de',
    u'Mej\xeda Juan Alberto Cosquillo'],
   'description': ['Cardiomyopathy, Dilated',
    'Cross-Sectional Studies',
    'Exercise Test',
    'Female',
    'Heart Transplantation',
    'Humans',
    'Intraoperative Complications',
    'Male',
    'Middle Aged',
    'Postoperative Complications',
    'Reference Values',
    'Retrospective Studies',
    'Tachycardia',
    'Time Factors',
    'Treatment Outcome',
    'Ventricular Dysfunction, Right'],
   'keyword': [],
   'lang': ['eng', 'por']}),
 ('14360229',
  {'auth': ['HOHORST H J'],
   'description': ['Antibodies',
    'Antigens',
    'Bacteria',
    'Blood Group Antigens',
    'Escherichia',
    'Humans',
    'O Antigens',
    'Salmonella'],
   'keyword': ['ANTIGENS AND ANTIBODIES',
    'BLOOD GROUPS',
    'ESCHERICHIA',
    'SALMONELLA'

### Part1: Community Detection

In [4]:
from itertools import combinations
import networkx as nx

def create_graph():
    '''
    return:
    a graph with author as node
    coauthorship as edge
    language used as attribute
    '''
    G = nx.Graph()
    for v in papers.itervalues():
        langs = v['lang']
        auths = v['auth']
        if len(auths) > 1:
            G.add_nodes_from(auths)
            for comb in combinations(auths, 2):
                for lang in langs:
                    G.add_edge(comb[0], comb[1], lang=lang)
        else:
            G.add_node(auths[0])
    print 'Number of nodes: {}'.format(G.number_of_nodes())
    print 'Number of edges: {}'.format(G.number_of_edges())
    return G

G_auth = create_graph()

Number of nodes: 476558
Number of edges: 1090012


In [5]:
from pprint import pprint

langList = {}
for v in papers.itervalues():
    for lang in v['lang']:
        if langList.get(lang) is None:
            langList[lang] = set([])
        for auth in v['auth']:
                langList[lang].add(auth)

# count the lanuage used in the papers
pprint(sorted([(k,len(v)) for (k,v) in langList.iteritems()], key=lambda x: -x[1])[:20])

[('eng', 397167),
 ('fre', 16274),
 ('ger', 14503),
 ('spa', 9819),
 ('rus', 9669),
 ('ita', 7879),
 ('pol', 6307),
 ('chi', 6161),
 ('jpn', 4757),
 ('por', 3675),
 ('cze', 2441),
 ('hun', 1546),
 ('rum', 1321),
 ('dut', 1199),
 ('bul', 842),
 ('und', 818),
 ('swe', 791),
 ('hrv', 690),
 ('ukr', 515),
 ('dan', 493)]


In [6]:
'''
Count the languages with the highest cluster coefficient. Since the group with less than 1000 entities
may have extreme stats, I would leave out the languages used less than 1000 times
'''
langs = {k:v for k,v in langList.iteritems() if len(v) > 1000}
cc = {k:nx.average_clustering(G_auth,v) for k,v in langs.iteritems()}
sorted(cc.iteritems(), key=lambda x: -x[1])[:10]

[('chi', 0.7984509515953556),
 ('rum', 0.7478133083091381),
 ('eng', 0.7253124238106683),
 ('por', 0.6679958821863696),
 ('spa', 0.6059905017850732),
 ('rus', 0.5849271128734129),
 ('pol', 0.5642259921085655),
 ('jpn', 0.5472025628319449),
 ('fre', 0.5434637970409696),
 ('hun', 0.5153554056161257)]

According to the numbers, there is no significant evidence of correlation between the size of the user group and the clustering coefficient. Considering the differences between Chinese/Rumanian and English, the culture of people using different language and the language distances, the explanation could be complex. 

### Part2: Co-occurence of Attributes

In [7]:
from collections import Counter

# count the ocurrances of descriptions and keywords
counter_descriptor = Counter()
counter_keyword = Counter()
for v in papers.itervalues():
    counter_descriptor.update(v['description'])
    counter_keyword.update(v['keyword'])

In [8]:
# take a look at the top descriptions
counter_descriptor.most_common(10)

[('Humans', 102582),
 ('Female', 37088),
 ('Male', 35074),
 ('Animals', 32425),
 ('Adult', 22722),
 ('Middle Aged', 18004),
 ('Aged', 12365),
 ('Adolescent', 10009),
 ('Child', 9782),
 ('Rats', 8858)]

In [9]:
# take a look at the top keywords
counter_keyword.most_common(10)

[('Population', 1826),
 ('Demographic Factors', 1745),
 ('Developing Countries', 1659),
 ('Population Dynamics', 1376),
 ('Economic Factors', 1205),
 ('Developed Countries', 1122),
 ('Research Methodology', 1026),
 ('Population Characteristics', 932),
 ('Asia', 780),
 ('EXPERIMENTAL LAB STUDY', 719)]

In [10]:
G_des = nx.Graph()
for v in papers.itervalues():
    for comb in combinations(v['description'], 2):
        G_des.add_node(comb[0])
        G_des.add_node(comb[1])
        G_des.add_edge(comb[0], comb[1])

In [11]:
from datetime import datetime

# calculate the average length of the shortest paths
# if you need more detailed timing and performance info, use cProfile module
t0 = datetime.now()

def average_shortest_path_length(G):
    length = 0
    for u,v in G.edges_iter():
        if u is not None and v is not None:
            spl = nx.shortest_path_length(G, u, v)
            length += spl
    n = len(G.edges())
    return length / float(n)

avg_length = average_shortest_path_length(G_des)
    
t1 = datetime.now()
print t1-t0

0:08:00.965552


In [12]:
avg_length

0.9968995591316296

In [13]:
# count the occurance of pairs of descriptions
counter_edge = Counter()
for v in papers.itervalues():
    # sort the list of description to make sure that there would not be tuples with inverse order
    sorted_des = sorted(v['description'])
    for e in combinations(sorted_des, 2):
        if e:
            counter_edge.update((e,))
counter_edge.most_common(10) 

[(('Female', 'Humans'), 32219),
 (('Humans', 'Male'), 28590),
 (('Adult', 'Humans'), 22653),
 (('Female', 'Male'), 21715),
 (('Humans', 'Middle Aged'), 17968),
 (('Adult', 'Female'), 15806),
 (('Adult', 'Male'), 15163),
 (('Male', 'Middle Aged'), 13452),
 (('Female', 'Middle Aged'), 13247),
 (('Aged', 'Humans'), 12310)]

In [14]:
from scipy.stats import chi2_contingency

# for each pair of description, run the chi square independence test

def calculate_chi2(T, c1, c2):
    table = {}
    for u,v in c2:
        crosstab = np.array([[0,0],[0,0]])
        val = c2[(u,v)]
        crosstab[1][1] += val
        crosstab[0][1] += (c1[u]-val)
        crosstab[1][0] += (c1[v]-val)
        crosstab[0][0] = T - crosstab[0][1] - crosstab[1][1] - crosstab[1][0]
        chisq = chi2_contingency(crosstab)[0]
        table[(u,v)] = chisq
    return table

T = len(papers)
chi2_table = calculate_chi2(T, counter_descriptor, counter_edge)

In [15]:
# take a look at the distribution of the chi2_table
print 'total number of pairs of desriptions: {}'.format(len(chi2_table))
pprint(np.histogram(chi2_table.values()))

p = 80
print '{0}% of the values fall into [0, {1}]'.format(p, np.percentile(chi2_table.values(), p))

total number of pairs of desriptions: 2429977
(array([2427178,    1812,     599,     187,      81,      46,      38,
            22,       5,       9]),
 array([  9.65653862e-11,   2.02875925e+04,   4.05751851e+04,
         6.08627776e+04,   8.11503701e+04,   1.01437963e+05,
         1.21725555e+05,   1.42013148e+05,   1.62300740e+05,
         1.82588333e+05,   2.02875925e+05]))
80% of the values fall into [0, 66.6193527632]


In [16]:
# take 67 as the threshold, select the pairs of descriptions with ch2 value greater than 67
selected_pairs = [k for k,v in chi2_table.iteritems() if v > 67]

# construct the selected graph for analysis
G_sel = nx.Graph()
for u,v in selected_pairs:
    G_sel.add_nodes_from([u,v])
    G_sel.add_edges_from([(u,v)])
print 'Number of nodes: {}'.format(len(G_sel.nodes()))
print 'Number of edges: {}'.format(len(G_sel.edges()))

Number of nodes: 22044
Number of edges: 484434


In [17]:
avg_length_sel = average_shortest_path_length(G_sel)
avg_length_sel

0.9999525219121697

The average shortest path length does not change a lot after the pairs of descriptions with low chi square values are removed - the connections between each pair of nodes(the authors) are already strong enough, which is not exactly what I have expected. On the other hand, having a stable stats after some of the nodes and edges are removed is a sign of a stable network, indicating some attributes of a small-world network.

### Part3: Link Prediction

In [18]:
def train_vali_split(G):
    '''
    parameter: the co-authorship graph
    return: training graph and validation graph
    split ratio: 7:3
    '''
    G_vali = nx.Graph()
    vali_nodes = [n for n in G.nodes_iter() if np.random.random() <= 0.3]
    G_vali = G.subgraph(vali_nodes)
    G.remove_nodes_from(vali_nodes)
    print 'Number of nodes in training graph: {}'.format(len(G.nodes()))
    print 'Number of edges in training graph: {}'.format(len(G.edges()))    
    print 'Number of nodes in validation graph: {}'.format(len(G_vali.nodes()))
    print 'Number of edges in validation graph: {}'.format(len(G_vali.edges()))
    return G, G_vali

G_train, G_vali = train_vali_split(G_auth)

Number of nodes in training graph: 333958
Number of edges in training graph: 535985
Number of nodes in validation graph: 142600
Number of edges in validation graph: 97827


#### Change of Split Method

It seems that although the nodes ratio is close to 7:3, it is not a good idea to use subgraph() to do the split, since the random selection of nodes would lead to a much less coupled graph wiht much fewer edges. This situation would lead to the underfit of the validation set, since it is less likely to have edges in the validation graph. 

Thankfully we have the original papers data, instead I'm going to construct the train/validation graph from scratch.

In [19]:
def split_from_origin(papers):
    '''
    parameter: dictionary of paper: authors
    return: training graph and validation graph
    using number of co-author(s) as weight
    split ratio: 7:3
    '''
    G_train = nx.Graph()
    G_vali = nx.Graph()
    for k,v in papers.iteritems():
        auths = v['auth']
        if np.random.random() <= 0.7:
            G_train.add_nodes_from(auths)
            for comb in combinations(auths, 2):
                G_train.add_edge(comb[0], comb[1], weight=len(auths))
        else:
            G_vali.add_nodes_from(auths)
            for comb in combinations(auths, 2):
                G_vali.add_edge(comb[0], comb[1], weight=len(auths))
    print 'Number of nodes in training graph: {}'.format(len(G_train.nodes()))
    print 'Number of edges in training graph: {}'.format(len(G_train.edges()))    
    print 'Number of nodes in validation graph: {}'.format(len(G_vali.nodes()))
    print 'Number of edges in validation graph: {}'.format(len(G_vali.edges()))
    return G_train, G_vali

G_train, G_vali = split_from_origin(papers)

Number of nodes in training graph: 354681
Number of edges in training graph: 773047
Number of nodes in validation graph: 169446
Number of edges in validation graph: 337784


In [46]:
# now the train/validation graphs look much better, let's get started!
import math

def get_data(G):
    
    N = len(G.nodes())
    print 'Number of nodes: {}'.format(N)
    
    data = {}
    data['label'] = {}
    flist = ['Jaccard', 'Cosine', 'Adar', 'common_neighbors', 'PageRank', 'kNN1', 'kNN2']
    for f in flist:
        data[f] = {}
    
    # personalized pagerank, setting the damping parameter as 0.8
    # return a dictionary of node: score
    pagerank = nx.pagerank(G, alpha=0.8, weight='weight')
    
    # since the possible edge set would be enourmous, we need to use previous pagerank
    # to filter the nodes with highest score
    N_high = N/200
    N_candidate = N_high * (N_high-1) / 2
    print 'Number of edges to calculate: {}'.format(N_candidate)
    
    nodes = dict(sorted(pagerank.iteritems(), key=lambda x: -x[1])[:N_high])
    k = 0
    K = 0
    for u,v in combinations(nodes.keys(), 2):
        
        un = [_ for _ in nx.all_neighbors(G, u)]
        vn = [_ for _ in nx.all_neighbors(G, v)]
        cn = set(un) & set(vn)
        union = set(un) | set(vn)
        un_len = len(un)
        vn_len = len(vn)
        cn_len = len(cn)
        
        data['common_neighbors'][(u,v)] = cn_len
        data['Jaccard'][(u,v)] = float(cn_len) / len(union)
        data['Cosine'][(u,v)] = float(cn_len) / (un_len * vn_len)
        
        if cn_len == 1 or cn_len == 0:
            data['Adar'][(u,v)] = 0
        else:
            data['Adar'][(u,v)] = 1.0 / math.log(cn_len)
        
        # the ways to calculate kNN weights are various, per se wa+wb, wa * wb
        # even deeper (2nd level or more) connections can be taken into account
        # for the sake of performance, I chose the 1st level calculation as weights
        data['kNN1'][(u,v)] = (1 / math.sqrt(1 + un_len)) + (1 / math.sqrt(1 + vn_len))
        data['kNN2'][(u,v)] = (1 / math.sqrt(1 + un_len)) * (1 / math.sqrt(1 + vn_len))
        data['PageRank'][(u,v)] = nodes[u] * nodes[v]
        
        if G.has_edge(u,v) is True:
            data['label'][(u,v)] = 1
        else:
            data['label'][(u,v)] = 0
        
        k += 1
        if k % (N_candidate/10) == 0:
            K += 10
            print 'Current progress: {}%'.format(K) 
    return data

In [47]:
data_train = get_data(G_train)

Number of nodes: 354681
Number of edges to calculate: 1570878
Current progress: 10%
Current progress: 20%
Current progress: 30%
Current progress: 40%
Current progress: 50%
Current progress: 60%
Current progress: 70%
Current progress: 80%
Current progress: 90%
Current progress: 100%


In [49]:
# just to ensure the index is correct, which is crucial to the problem
df_train = pd.DataFrame(data_train)
df_train.head()

Adar  Cosine  Jaccard      PageRank  \
(ARNOULD P, Abayomi A)              0       0        0  9.442268e-11   
(ARNOULD P, Abe H)                  0       0        0  1.279890e-10   
(ARNOULD P, Abe K)                  0       0        0  1.612973e-10   
(ARNOULD P, Abe T)                  0       0        0  1.474687e-10   
(ARNOULD P, Abolmaesumi Purang)     0       0        0  1.243724e-10   

                                 common_neighbors      kNN1      kNN2  label  
(ARNOULD P, Abayomi A)                          0  0.634845  0.100504      0  
(ARNOULD P, Abe H)                              0  0.462433  0.043033      0  
(ARNOULD P, Abe K)                              0  0.457368  0.041345      0  
(ARNOULD P, Abe T)                              0  0.460333  0.042333      0  
(ARNOULD P, Abolmaesumi Purang)                 0  0.591532  0.086066      0

In [50]:
data_vali = get_data(G_vali)

Number of nodes: 169446
Number of edges to calculate: 358281
Current progress: 10%
Current progress: 20%
Current progress: 30%
Current progress: 40%
Current progress: 50%
Current progress: 60%
Current progress: 70%
Current progress: 80%
Current progress: 90%
Current progress: 100%


In [51]:
df_vali = pd.DataFrame(data_vali)
df_vali.head()

Adar  Cosine  Jaccard      PageRank  common_neighbors  \
(Abe T, Akbulut H)     0       0        0  5.528048e-10                 0   
(Abe T, BONNAL J)      0       0        0  5.470430e-10                 0   
(Abe T, Berg T)        0       0        0  5.673243e-10                 0   
(Abe T, Bird E D)      0       0        0  5.742763e-10                 0   
(Abe T, Bloom S R)     0       0        0  9.635237e-10                 0   

                        kNN1      kNN2  label  
(Abe T, Akbulut H)  0.387298  0.033333      0  
(Abe T, BONNAL J)   0.430611  0.038925      0  
(Abe T, Berg T)     0.406450  0.035806      0  
(Abe T, Bird E D)   0.347317  0.028172      0  
(Abe T, Bloom S R)  0.321550  0.024845      0

In [57]:
df_train.to_csv('train.csv', index=False)
df_vali.to_csv('validate.csv', index=False)

In [1]:
import h2o
h2o.init()

------------------------------  ------------------------------------
H2O cluster uptime:             7 minutes 53 seconds 98 milliseconds
H2O cluster version:            3.8.1.3
H2O cluster name:               light
H2O cluster total nodes:        1
H2O cluster total free memory:  1.56 GB
H2O cluster total cores:        4
H2O cluster allowed cores:      4
H2O cluster healthy:            True
H2O Connection ip:              127.0.0.1
H2O Connection port:            54321
H2O Connection proxy:
Python Version:                 2.7.11
------------------------------  ------------------------------------

In [2]:
import os

basedir = os.getcwd()
fr_train = h2o.import_file(path=basedir+"/train.csv")
fr_train.head(5)


Parse Progress: [##################################################] 100%


Adar Cosine Jaccard PageRank common_neighbors kNN1 kNN2 label 
 0 0 0 9.44227e-11 0 0.634845 0.100504 0 
 0 0 0 1.27989e-10 0 0.462433 0.0430331 0 
 0 0 0 1.61297e-10 0 0.457368 0.0413449 0 
 0 0 0 1.47469e-10 0 0.460333 0.0423334 0 
 0 0 0 1.24372e-10 0 0.591532 0.0860663 0

In [3]:
fr_vali = h2o.import_file(path=basedir+"/validate.csv")
fr_vali.head(5)


Parse Progress: [##################################################] 100%


Adar Cosine Jaccard PageRank common_neighbors kNN1 kNN2 label 
 0 0 0 5.52805e-10 0 0.387298 0.0333333 0 
 0 0 0 5.47043e-10 0 0.430611 0.0389249 0 
 0 0 0 5.67324e-10 0 0.40645 0.0358057 0 
 0 0 0 5.74276e-10 0 0.347317 0.0281718 0 
 0 0 0 9.63524e-10 0 0.32155 0.0248452 0

In [5]:
fr_train['label'] = fr_train['label'].asfactor()
fr_vali['label'] = fr_vali['label'].asfactor()

from h2o.estimators.gbm import H2OGradientBoostingEstimator


gbm = H2OGradientBoostingEstimator(distribution='bernoulli',
                                   ntrees=100,
                                   max_depth=5,
                                   balance_classes=True,
                                   learn_rate=0.01)

gbm.train(x=fr_train.columns[:-1],
          y='label', 
          training_frame=fr_train,
          validation_frame=fr_vali)

gbm.show()


gbm Model Build Progress: [##################################################] 100%
Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Method
Model Key:  GBM_model_python_1458712020197_2

Model Summary: 


number_of_trees    model_size_in_bytes    min_depth    max_depth    mean_depth    min_leaves    max_leaves    mean_leaves
--  -----------------  ---------------------  -----------  -----------  ------------  ------------  ------------  -------------
    100                40353                  5            5            5             28            32            29.36



ModelMetricsBinomial: gbm
** Reported on train data. **

MSE: 0.491430863128
R^2: -0.965723452513
LogLoss: 2.38680414851
AUC: 0.998721217075
Gini: 0.997442434149

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.00238006530015: 


0            1            Error    Rate
-----  -----------  -----------  -------  -------------------
0      1.55148e+06  16225        0.0103   (16225.0/1567703.0)
1      9875         1.55783e+06  0.0063   (9875.0/1567704.0)
Total  1.56135e+06  1.57405e+06  0.0083   (26100.0/3135407.0)


Maximum Metrics: Maximum metrics at their respective thresholds



metric                      threshold    value     idx
--------------------------  -----------  --------  -----
max f1                      0.00238007   0.991693  246
max f2                      0.00190171   0.993453  273
max f0point5                0.00350894   0.99142   185
max accuracy                0.00238007   0.991676  246
max precision               0.00891278   0.999914  0
max recall                  0.000462833  1         394
max specificity             0.00891278   0.999955  0
max absolute_MCC            0.00238007   0.98336   246
max min_per_class_accuracy  0.00259576   0.990754  232


Gains/Lift Table: Avg response rate:  0.20 %



group    cumulative_data_fraction    lower_threshold    lift        cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  ----------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0100994                   0.00289197         97.8303     97.8303            0.197731         0.197731                    0.988031        0.988031                   9683.03   9683.03
    2        0.0200187                   0.000934892        0.762056    49.7329            0.00154024       0.100518                    0.00755906      0.995591                   -23.7944  4873.29
    3        0.0331382                   0.000498151        0.0720216   30.0721            0.000145567      0.0607807                   0.000944882     0.996535                   -92.7978  2907.21
    4        0.0439003                   0.000497411        0           22.7               0                0.0458803                   0               0.996535                   -100      2170
    5        0.0526425                   0.000482719        0           18.9302            0                0.0382611                   0               0.996535                   -100      1793.02
    6        0.120419                    0.000466498        0.0278824   8.29128            5.6355e-05       0.016758                    0.00188976      0.998425                   -97.2118  729.128
    7        0.151876                    0.000463814        0           6.57397            0                0.0132871                   0               0.998425                   -100      557.397
    8        0.319027                    0.000463125        0.00188428  3.13058            3.80844e-06      0.00632741                  0.000314961     0.99874                    -99.8116  213.058
    9        0.520509                    0.000462833        0.00625289  1.9212             1.26381e-05      0.00388305                  0.00125984      1                          -99.3747  92.1197
    10       0.725633                    0.00046196         0           1.37811            0                0.00278538                  0               1                          -100      37.8107
    11       0.998346                    0.000460337        0           1.00166            0                0.00202451                  0               1                          -100      0.165659
    12       1                           0.000438946        0           1                  0                0.00202116                  0               1                          -100      0



ModelMetricsBinomial: gbm
** Reported on validation data. **

MSE: 0.00406798125534
R^2: 0.0131232075352
LogLoss: 0.0203166520275
AUC: 0.997963304709
Gini: 0.995926609418

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.00882852677271: 


0       1     Error    Rate
-----  ------  ----  -------  ----------------
0      356652  146   0.0004   (146.0/356798.0)
1      144     1339  0.0971   (144.0/1483.0)
Total  356796  1485  0.0008   (290.0/358281.0)


Maximum Metrics: Maximum metrics at their respective thresholds



metric                      threshold    value     idx
--------------------------  -----------  --------  -----
max f1                      0.00882853   0.902291  8
max f2                      0.00852306   0.916819  22
max f0point5                0.00884966   0.908204  5
max accuracy                0.00882853   0.999191  8
max precision               0.00891278   0.934075  0
max recall                  0.000463814  1         197
max specificity             0.00891278   0.999767  0
max absolute_MCC            0.00882853   0.901885  8
max min_per_class_accuracy  0.00473891   0.988537  79


Gains/Lift Table: Avg response rate:  0.41 %



group    cumulative_data_fraction    lower_threshold    lift        cumulative_lift    response_rate    cumulative_response_rate    capture_rate    cumulative_capture_rate    gain      cumulative_gain
--  -------  --------------------------  -----------------  ----------  -----------------  ---------------  --------------------------  --------------  -------------------------  --------  -----------------
    1        0.0100284                   0.00557003         97.9009     97.9009            0.405232         0.405232                    0.981794        0.981794                   9690.09   9690.09
    2        0.0201015                   0.00452636         0.803299    49.2443            0.00332502       0.203832                    0.00809171      0.989885                   -19.6701  4824.43
    3        0.0459751                   0.00106183         0.234555    21.6629            0.000970874      0.0896673                   0.00606878      0.995954                   -76.5445  2066.29
    4        0.0514959                   0.00105983         0           19.3405            0                0.0800542                   0               0.995954                   -100      1834.05
    5        0.146851                    0.00105328         0           6.78206            0                0.0280724                   0               0.995954                   -100      578.206
    6        0.179013                    0.000951127        0           5.56358            0                0.0230288                   0               0.995954                   -100      456.358
    7        0.200867                    0.000482719        0.0617093   4.96498            0.000255428      0.0205511                   0.00134862      0.997303                   -93.8291  396.498
    8        0.525426                    0.000466885        0.00415524  1.90065            1.71994e-05      0.0078672                   0.00134862      0.998651                   -99.5845  90.0652
    9        0.8418                      0.000464367        0.00213136  1.18713            8.82215e-06      0.00491378                  0.000674309     0.999326                   -99.7869  18.7129
    10       0.913897                    0.000463814        0.00935279  1.09421            3.87132e-05      0.00452918                  0.000674309     1                          -99.0647  9.4215
    11       1                           0.000457076        0           1                  0                0.00413921                  0               1                          -100      0



Scoring History: 


timestamp            duration          number_of_trees    training_MSE    training_logloss    training_AUC    training_lift    training_classification_error    validation_MSE    validation_logloss    validation_AUC    validation_lift    validation_classification_error
--  -------------------  ----------------  -----------------  --------------  ------------------  --------------  ---------------  -------------------------------  ----------------  --------------------  ----------------  -----------------  ---------------------------------
    2016-03-22 22:56:21  0.058 sec         0                  0.497983        3.10305             0.5             1                0.5                              0.00412656        0.0276948             0.5               1                  0.995861
    2016-03-22 22:56:27  6.351 sec         1                  0.497944        3.09339             0.997489        72.0553          0.0106088                        0.0041261         0.0275812             0.995089          72.7053            0.000862452
    2016-03-22 22:56:41  19.677 sec        9                  0.497615        3.01931             0.997512        93.9624          0.0104136                        0.0041225         0.0267308             0.996209          70.8428            0.000862452
    2016-03-22 22:57:14  52.389 sec        28                 0.496724        2.86155             0.997598        95.8448          0.00990812                       0.00411387        0.0250174             0.997381          65.5955            0.000848496
    2016-03-22 22:58:31  2 min 10.060 sec  75                 0.493685        2.53715             0.99869         98.4878          0.00852648                       0.00408718        0.0217288             0.997823          97.8192            0.000809421
    2016-03-22 22:59:33  3 min 11.958 sec  100                0.491431        2.3868              0.998721        97.8303          0.00832428                       0.00406798        0.0203167             0.997963          97.9009            0.000809421


Variable Importances: 


variable          relative_importance    scaled_importance    percentage
----------------  ---------------------  -------------------  ------------
Jaccard           3.26457e+07            1                    0.98713
Adar              241428                 0.00739541           0.00730023
PageRank          70037.7                0.00214539           0.00211778
kNN1              57721.3                0.00176812           0.00174536
Cosine            32514.1                0.00099597           0.000983152
kNN2              23931.6                0.00073307           0.000723636
common_neighbors  0                      0                    0

#### Link prediction result:

Without feature engineering and grid search, the model is able to provide prediction with accuracy of 99.92% and AUC of 99.79% which proves that the choices of the features are correct! Even facing the sparsity of social network, this model ahieve its goal with basic attributes calculations and even 'out-of-date' PageRank algorithm. On this dataset, I would say the link prediction problem is solved.